### ISSUE TITLE
Explain why borrows can't be held across yield point in async blocks/functions

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust#78938]
Consider the following code:

```rust
use std::sync::Arc;
use tokio::runtime::Runtime; // 0.3.1

async fn f() {
    let room_ref = Arc::new(Vec::new());

    let gameloop_handle = Runtime::new().unwrap().spawn(async {
        game_loop(Arc::clone(&room_ref))
    });
    gameloop_handle.await;
}

fn game_loop(v: Arc<Vec<usize>>) {}
```

The error message has a helpful hint:

```
help: to force the async block to take ownership of `room_ref` (and any other referenced variables), use the `move` keyword
  |
7 |     let gameloop_handle = Runtime::new().unwrap().spawn(async move {
8 |         game_loop(Arc::clone(&room_ref))
9 |     });
```

But it doesn't explain very well _why_ `move` is necessary:

```
error[E0373]: async block may outlive the current function, but it borrows `room_ref`, which is owned by the current function
 --> src/lib.rs:7:63
  |
7 |       let gameloop_handle = Runtime::new().unwrap().spawn(async {
  |  _______________________________________________________________^
8 | |         game_loop(Arc::clone(&room_ref))
  | |                               -------- `room_ref` is borrowed here
9 | |     });
  | |_____^ may outlive borrowed value `room_ref`
  |
note: function requires argument type to outlive `'static`
 --> src/lib.rs:7:57
  |
7 |       let gameloop_handle = Runtime::new().unwrap().spawn(async {
  |  _________________________________________________________^
8 | |         game_loop(Arc::clone(&room_ref))
9 | |     });
  | |_____^
```

In particular, 'async block may outlive the current function' makes no sense without a very good mental model of async: the future is `await`ed within the current function, so of course it can't outlive it. The thing to realize here is that even though it's _written_ as one function, it's actually two: one executed before the yield point, and one after, and the stack space for the first function goes away when you call `await`.

It would be nice to instead say something like

```
note: borrows cannot be held across a yield point, because the stack space of the current function is not preserved
help: see https://rust-lang.github.io/async-book/03_async_await/01_chapter.html#awaiting-on-a-multithreaded-executor for more information
```

**Comment by jyn514** [comment]
cc @estebank

**Comment by jyn514** [comment]
Hmm, reading that link, it looks like this issue is specific to multi-threaded executors and this would be fine without a `'static` bound on `spawn`? I'm not sure how much help the compiler can be then ...

**Comment by SkiFire13** [comment]
This isn't specific to async, it also happens with `std::thread::spawn` requiring the closure to be `'static` even if we later call `JoinHandle::join` in the same function.

**Comment by tmandry** [comment]
What about something like

```
note: async blocks are not executed immediately and either must take a reference or ownership of outside variables they use
help: to force ...
```

We could substitute "closures" in the non-async case as well, if it'd be helpful.

**Comment by tmandry** [comment]
Happy to help anyone who wants to implement this.

**Comment by 1000teslas** [comment]
Hi, I want to start contributing to Rust. Would this be a good first issue?

**Comment by jyn514** [comment]
@1000teslas that'd be great! See https://rustc-dev-guide.rust-lang.org/ for instructions on getting started, and for your change I'd start by finding where the current error is emitted and looking around there.

**Comment by 1000teslas** [comment]
I think the current error message is emitted from `report_escaping_closure_capture` in `compiler/rustc_mir/src/borrow_check/diagnostics/conflict_errors.rs`. However, I'm not sure what the appropriate method for adding those new help/note messages. Would the `help` and `note` messages from `DiagnosticBuilder` be the right ones?

**Comment by jyn514** [comment]
@1000teslas yes, those are right (or at least close enough that you can fix any issues during review).

**Comment by 1000teslas** [comment]
I have tried implementing the error message for async block here: [https://github.com/rust-lang/rust/pull/80614]. I am not sure what errors an async function would have. For example, I tried messing around here [https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=774caaba98a35c9ba2d07a9f3396473a] with borrowing past an await point in an async function, but it seems to compile.

I am not sure how to keep lines under 100 characters when mentioning urls in an error message. Also, for some reason, `./x.py fmt` doesn't seem to fix up the formatting of my files.
